### PR TITLE
fix: add FlyteValueException handling and clean up exit_handler calls

### PR DESCRIPTION
**Pull request by vincent0426**
## Why are the changes needed?

Currently, if a task returns a value but does not specify a return type, the return value will be empty instead of raising an error.

It is expected to raise an error if a function returns a value but no return type is specified.

This should throw an error, but it doesn't 

`pyflyte run example.py greeting_length --greeting="hello"`
```python
@task()
def greeting_length(greeting: str):
    """A task the counts the length of a greeting."""
    return len(greeting)
```

Screenshot 2025-01-05 at 8 57 25 PM

## What changes were proposed in this pull request?

Add conditions check to throw an error if `native_outputs` have value but no output interface is specified

### Setup process

`pyflyte run example.py greeting_length --greeting="hello"`
```python
@task()
def greeting_length(greeting: str):
    """A task the counts the length of a greeting."""
    return len(greeting)
```

### Screenshots
Screenshot 2025-01-05 at 8 58 41 PM

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [X] All new and existing tests passed.
- [X] All commits are signed-off.

### CODE REVIEW

File: plugins/flytekit-flyteinteractive/flytekitplugins/flyteinteractive/jupyter_lib/decorator.py [hunk 1]
```diff
@@ -168,7 +168,7 @@ def execute(self, *args, **kwargs):
 
         write_example_notebook(task_function=self.task_function, notebook_dir=self.notebook_dir)
 
-        return exit_handler(
+        exit_handler(
```
Comment (vincent0426) [inline plugins/flytekit-flyteinteractive/flytekitplugins/flyteinteractive/jupyter_lib/decorator.py:171]: Remove the `return` for both Jupyter and VS Code. I'm not sure if this will cause any side effects. If not removed, native_output will receive the mock function, which will then throw an error as the output interface is empty.

Comment (pingsutw) [reply plugins/flytekit-flyteinteractive/flytekitplugins/flyteinteractive/jupyter_lib/decorator.py:171]: We cannot remove `return` here. Where did you see the error?

Comment (vincent0426) [reply plugins/flytekit-flyteinteractive/flytekitplugins/flyteinteractive/jupyter_lib/decorator.py:171]: Hi @pingsutw 

https://github.com/flyteorg/flytekit/blob/ea17a4d3c368d4aff4db3631ec53f78805955cfc/flytekit/core/base_task.py#L779-L783


the following tests will fail as they would return the mock handler but the output interface expects to have 0 outputs

`pytest -sv tests/flytekit/unit/interactive/test_flyteinteractive_vscode.py::test_vscode_remote_execution`

https://github.com/flyteorg/flytekit/blob/9c2caa10e8b335f2e684ebd220dd47a8d2ec0f92/tests/flytekit/unit/interactive/test_flyteinteractive_vscode.py#L139-L142

Screenshot 2025-01-06 at 2 33 33 PM

this one should fail, too
https://github.com/flyteorg/flytekit/blob/9c2caa10e8b335f2e684ebd220dd47a8d2ec0f92/plugins/flytekit-flyteinteractive/tests/test_flyteinteractive_jupyter.py#L38-L41

Comment (pingsutw) [reply plugins/flytekit-flyteinteractive/flytekitplugins/flyteinteractive/jupyter_lib/decorator.py:171]: ah, you can add return type to that function.
```
@task
    @vscode
    def t() -> str:
        return "hello"
```

Comment (vincent0426) [reply plugins/flytekit-flyteinteractive/flytekitplugins/flyteinteractive/jupyter_lib/decorator.py:171]: Adding `-> str` might require modifying the return value of exit_handler() as t() may be overridden by the decorator?

```python
mock.patch(
        "flytekit.interactive.vscode_lib.decorator.exit_handler",
        return_value="Hello, world!",
```

However, this will further lead to other errors, which I don't think is solvable as @vscode may not always return exit_handler(). Thanks.
 
Screenshot 2025-01-06 at 3 38 48 PM

### full error log

vscode_patches = (<MagicMock name='Process' id='5161833280'>, <MagicMock name='prepare_interactive_python' id='5161902128'>, <MagicMock...10240'>, <MagicMock name='signal' id='5162650272'>, <MagicMock name='prepare_resume_task_python' id='5162654928'>, ...)
mock_local_execution = <MagicMock name='is_local_execution' id='5162662848'>

    def test_vscode_local_execution(vscode_patches, mock_local_execution):
        (
            mock_process,
            mock_prepare_interactive_python,
            mock_exit_handler,
            mock_download_vscode,
            mock_signal,
            mock_prepare_resume_task_python,
            mock_prepare_launch_json,
        ) = vscode_patches
    
        @task
        @vscode
        def t() -> str:
            return
    
        @workflow
        def wf():
            t()
    
>       wf()

tests/flytekit/unit/interactive/test_flyteinteractive_vscode.py:149: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
flytekit/core/workflow.py:308: in __call__
    raise exc
flytekit/core/workflow.py:301: in __call__
    return flyte_entity_call_handler(self, *args, **input_kwargs)
flytekit/core/promise.py:1526: in flyte_entity_call_handler
    return cast(LocallyExecutable, entity).local_execute(ctx, **kwargs)
flytekit/core/workflow.py:327: in local_execute
    function_outputs = self.execute(**kwargs_literals)
flytekit/core/workflow.py:838: in execute
    return self._workflow_function(**kwargs)
tests/flytekit/unit/interactive/test_flyteinteractive_vscode.py:147: in wf
    t()
flytekit/core/base_task.py:365: in __call__
    return flyte_entity_call_handler(self, *args, **kwargs)  # type: ignore
flytekit/core/promise.py:1526: in flyte_entity_call_handler
    return cast(LocallyExecutable, entity).local_execute(ctx, **kwargs)
flytekit/core/base_task.py:345: in local_execute
    outputs_literal_map = self.sandbox_execute(ctx, input_literal_map)
flytekit/core/base_task.py:422: in sandbox_execute
    return self.dispatch_execute(ctx, input_literal_map)
flytekit/core/base_task.py:787: in dispatch_execute
    literals_map, native_outputs_as_map = run_sync(
flytekit/utils/asyn.py:106: in run_sync
    return self._runner_map[name].run(coro)
flytekit/utils/asyn.py:85: in run
    res = fut.result(None)
/opt/homebrew/Cellar/python@3.12/3.12.8/Frameworks/Python.framework/Versions/3.12/lib/python3.12/concurrent/futures/_base.py:456: in result
    return self.__get_result()
/opt/homebrew/Cellar/python@3.12/3.12.8/Frameworks/Python.framework/Versions/3.12/lib/python3.12/concurrent/futures/_base.py:401: in __get_result
    raise self._exception
flytekit/core/base_task.py:656: in _output_to_literal_map
    raise e
flytekit/core/type_engine.py:1389: in async_to_literal
    cls.to_literal_checks(python_val, python_type, expected)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'flytekit.core.type_engine.TypeEngine'>, python_val = None, python_type = <class 'str'>
expected = Flyte Serialized object (LiteralType):
  simple: 3

    @classmethod
    def to_literal_checks(cls, python_val: typing.Any, python_type: Type[T], expected: LiteralType):
        from flytekit.core.promise import VoidPromise
    
        if isinstance(python_val, VoidPromise):
            raise AssertionError(
                f"Outputs of a non-output producing task {python_val.task_name} cannot be passed to another task."
            )
        if isinstance(python_val, tuple):
            raise AssertionError(
                "Tuples are not a supported type for individual values in Flyte - got a tuple -"
                f" {python_val}. If using named tuple in an inner task, please, de-reference the"
                "actual attribute that you want to use. For example, in NamedTuple('OP', x=int) then"
                "return v.x, instead of v, even if this has a single element"
            )
        if (
            (python_val is None and python_type != type(None))
            and expected
            and expected.union_type is None
            and python_type is not Any
        ):
>           raise TypeTransformerFailedError(f"Python value cannot be None, expected {python_type}/{expected}")
E           flytekit.core.type_engine.TypeTransformerFailedError: Failed to convert outputs of task '__channelexec__.test_flyteinteractive_vscode.t' at position 0.
E           Failed to convert type <class 'NoneType'> to type <class 'str'>.
E           Error Message: Python value cannot be None, expected <class 'str'>/Flyte Serialized object (LiteralType):
E             simple: 3.

flytekit/core/type_engine.py:1330: TypeTransformerFailedError

Comment (Mecoli1219) [reply plugins/flytekit-flyteinteractive/flytekitplugins/flyteinteractive/jupyter_lib/decorator.py:171]: Why not just mock the return value to None in the unit test?

**Comment by Mecoli1219** [comment]
@vincent0426 Can you add some unit tests to make sure running the task/workflow without specifying output while returning a non-None value would throw an error? Also, running with `--remote` flag should also throw an error, but I think you can fix it in another PR.
